stock_demand_estimate/wizards/stock_demand_estimate_wizard.py

```diff
@@ -0,0 +1,251 @@
+# Copyright 2016 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# Copyright 2016 Aleph Objects, Inc. (https://www.alephobjects.com/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import UserError, ValidationError
+
+
+class StockDemandEstimateSheet(models.TransientModel):
+    _name = 'stock.demand.estimate.sheet'
+    _description = 'Stock Demand Estimate Sheet'
+
+    date_start = fields.Date(
+        string="Date From",
+        readonly=True,
+    )
+    date_end = fields.Date(
+        string="Date to",
+        readonly=True,
+    )
+    date_range_type_id = fields.Many2one(
+        string='Date Range Type',
+        comodel_name='date.range.type',
+        readonly=True,
+    )
+    location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Location",
+        readonly=True,
+    )
+    line_ids = fields.Many2many(
+        string="Estimates",
+        comodel_name='stock.demand.estimate.sheet.line',
+        relation='stock_demand_estimate_line_rel',
+    )
+
+    @api.onchange('date_start', 'date_end', 'date_range_type_id',)
+    def _onchange_dates(self):
+        if not all([self.date_start, self.date_end, self.date_range_type_id]):
+            return
+        ranges = self._get_ranges()
+        product_ids = self.env.context.get('product_ids', [])
+        lines = []
+        for product in self.env['product.product'].browse(product_ids):
+            for _range in ranges:
+                estimate = self.env['stock.demand.estimate'].search([
+                    ('product_id', '=', product.id),
+                    ('date_range_id', '=', _range.id),
+                    ('location_id', '=', self.location_id.id)
+                ], limit=1)
+                if estimate:
+                    uom_id = estimate.product_uom.id
+                    uom_qty = estimate.product_uom_qty
+                    estimate_id = estimate.id
+                else:
+                    uom_id = product.uom_id.id
+                    uom_qty = 0.0
+                    estimate_id = None
+                lines.append((0, 0, self._get_default_estimate_line(
+                    _range,
+                    product,
+                    uom_id,
+                    uom_qty,
+                    estimate_id=estimate_id,
+                )))
+        self.write({'line_ids': lines})
+
+    def _get_ranges(self):
+        domain = [
+            ('type_id', '=', self.date_range_type_id.id), '|', '&',
+            ('date_start', '>=', self.date_start),
+            ('date_start', '<=', self.date_end),
+            '&',
+            ('date_end', '>=', self.date_start),
+            ('date_end', '<=', self.date_end),
+        ]
+        ranges = self.env['date.range'].search(domain)
```

Comment:
@mpanarin Is it possible to avoid a search ?

Maybe use the expression.OR() on both domains (this helps readability)